apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/ClientPool.cs

```diff
@@ -85,53 +108,182 @@ public override int GetHashCode()
-            public async Task<SpannerClient> AcquireClientFromEntryAsync()
+            public CredentialClientPool(ClientCredentialKey key) => _key = key;
+
+            //Returns the total sum of reference counts.
+            internal int DumpCredentialPoolContents(StringBuilder stringBuilder)
+            {
+                lock (_sync)
+                {
+                    int referenceCountTotal = 0;
+                    int i = 0;
+                    foreach (var item in _clientPriorityList.GetSnapshot())
+                    {
+                        stringBuilder.AppendLine($"  {i}:{item}");
+                        referenceCountTotal += item.RefCount;
+                        i++;
+                    }
+                    return referenceCountTotal;
+                }
+            }
+
+            public Task<SpannerClient> AcquireClientAsync()
             {
-                if (_client == null)
+                Task<SpannerClient> result;
+
+                lock (_sync)
                 {
-                    Logger.Debug(() => "Creating a new SpannerClient.");
-                    var endpoint = _key.Endpoint ?? SpannerClient.DefaultEndpoint;
-                    if (_key.Credential != null)
+                    //first ensure that the pool is of the correct size.
+                    while (_clientPriorityList.Count > SpannerOptions.Instance.MaximumGrpcChannels)
                     {
-                        lock (_sync)
-                        {
-                            if (_client == null)
-                            {
-                                var channel = new Channel(
-                                    endpoint.Host,
-                                    endpoint.Port,
-                                    _key.Credential.ToChannelCredentials());
-                                _client = SpannerClient.Create(channel);
-                            }
-                        }
+                        _clientPriorityList.RemoveLast();
                     }
-                    else
+                    while (_clientPriorityList.Count < SpannerOptions.Instance.MaximumGrpcChannels)
                     {
-                        lock (_sync)
-                        {
-                            if (_creationTask == null || _creationTask.IsFaulted)
-                            {
-                                _creationTask = SpannerClient.CreateAsync(endpoint);
-                            }
-                        }
-                        // await needs to be done outside of the monitor.
-                        _client = await _creationTask.ConfigureAwait(false);
+                        var newEntry = new SpannerClientCreator(_key);
+                        _clientPriorityList.Add(newEntry);
+                    }
+
+                    //now grab the first item in the sorted list, increment refcnt, re-sort and return.
+                    result = _clientPriorityList.GetTop().AcquireClientAsync();
```

Comment:
Perhaps add a comment to the effect that AcquireClientAsync automatically resorts due to informing the priority list of its own change? (I was trying to work out how that happened.)